xds/src/main/java/io/grpc/xds/Filter.java

```diff
@@ -0,0 +1,128 @@
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.protobuf.Any;
+import io.grpc.ClientInterceptor;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.ServerInterceptor;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.Nullable;
+
+/**
+ * Defines the parsing functionality of an HTTP filter. A Filter may optionally implement either
+ * {@link ClientInterceptorBuilder} or {@link ServerInterceptorBuilder} or both, indicating it is
+ * capable of working on the client side or server side or both, respectively.
+ */
+interface Filter {
+
+  /**
+   * The proto message types supported by this filter. A filter will be registered by each of its
+   * supported message types.
+   */
+  String[] typeUrls();
+
+  /**
+   * Parses the top-level filter config from raw proto message.
+   */
+  StructOrError<? extends FilterConfig> parseFilterConfig(Any rawProtoMessage);
+
+  /** Parses the per-filter override filter config from raw proto message. */
+  StructOrError<? extends FilterConfig> parseFilterConfigOverride(Any rawProtoMessage);
+
+  /** Represents an opaque data structure holding configuration for a filter. */
+  interface FilterConfig {
+    String typeUrl();
+  }
+
+  /** Uses the FilterConfigs produced above to produce an HTTP filter interceptor for clients. */
+  interface ClientInterceptorBuilder {
+    @Nullable
+    ClientInterceptor buildClientInterceptor(
+        FilterConfig config, @Nullable FilterConfig overrideConfig, PickSubchannelArgs args,
+        ScheduledExecutorService scheduler);
+  }
+
+  // Server side filters are not currently supported, but this interface is defined for clarity.
+  interface ServerInterceptorBuilder {
+    ServerInterceptor buildServerInterceptor(
+        FilterConfig config, @Nullable FilterConfig overrideConfig);
+  }
+
+  final class StructOrError<T> {
+    /**
+     * Returns a {@link StructOrError} for the successfully converted data object.
+     */
+    static <T> StructOrError<T> fromStruct(T struct) {
+      return new StructOrError<>(struct);
+    }
+
+    /**
+     * Returns a {@link StructOrError} for the failure to convert the data object.
+     */
+    static <T> StructOrError<T> fromError(String errorDetail) {
+      return new StructOrError<>(errorDetail);
+    }
+
+    final String errorDetail;
+    final T struct;
+
+    private StructOrError(T struct) {
+      this.struct = checkNotNull(struct, "struct");
+      this.errorDetail = null;
+    }
+
+    private StructOrError(String errorDetail) {
+      this.struct = null;
+      this.errorDetail = checkNotNull(errorDetail, "errorDetail");
+    }
+  }
+
+  final class Registry {
+    static Registry GLOBAL_REGISTRY =
+        newRegistry().register(FaultFilter.INSTANCE, RouterFilter.INSTANCE);
+
+    private final Map<String, Filter> map = new HashMap<>();
```

Comment:
nit: `map` is a bad name. `filters`/`supportedFilters` etc are better.

Reply:
Renamed.